rtmp-stream.go

```diff
@@ -98,6 +114,67 @@ func (f *RtmpStreamFactory) startStream(conn rtmp.ClientConn, streamName string)
 	}
 }
 
+func (f *RtmpStreamFactory) ParseURLArgument(urlArg string) (string, []*url.URL, error) {
+	var host string
+	var urls []*url.URL
+	var r = regexp.MustCompile(`{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}`) // {{123 456}}
+
+	var regexString = "'{{(?P<min>[1-9][0-9]*) (?P<max>[1-9][0-9]*)}}'"
+	var regexInfo = fmt.Sprintf("Use regex that matches patter %v", regexString)
+
+	if r.MatchString(urlArg) { // URL is a template.
+		log.Infoln("Processing template URL %v with regex matching. (Used regex: '%v')", urlArg, regexString)
+		match := r.FindStringSubmatch(urlArg)
+		min, er := strconv.Atoi(match[1])
+		if er != nil {
+			return "", nil, fmt.Errorf("Failed to parse minimal value from url %v. %v: %v", urlArg, regexInfo, er)
+		}
+		max, er := strconv.Atoi(match[2])
+		if er != nil {
+			return "", nil, fmt.Errorf("Failed to parse maximal value from url %v. %v: %v", urlArg, regexInfo, er)
+		}
+		if min > max {
+			return "", nil, fmt.Errorf("Minimal value cannot be greater than maximal value in url %v. %v.", urlArg, regexInfo)
+		}
+		if resultHost, resultURLs, er := f.generateURLs(urlArg, match[0], min, max); er == nil {
+			host = resultHost
+			urls = append(urls, resultURLs...)
+		} else {
+			return "", nil, fmt.Errorf("URL generation based on template URL %v failed. %v: %v", urlArg, regexInfo, er)
+		}
+	} else { // No matching regex expression found in url argument. URL is not a template. Returning it as it is.
+		log.Infoln("URL is not a template. Parsing URL %v without regex matching. (Used regex: '%v')", urlArg, regexString)
+		parsedURL, err := url.Parse(urlArg)
+		if err != nil {
+			return "", nil, fmt.Errorf("Failed to parse URL (%v): %v", urlArg, err)
+		}
+		host = parsedURL.Host
+		urls = append(urls, parsedURL)
+	}
+
+	return host, urls, nil
+}
+
+func (f *RtmpStreamFactory) generateURLs(urlArg string, toReplace string, min int, max int) (string, []*url.URL, error) {
+	var host string
+	var urls []*url.URL
+
+	if strings.Contains(urlArg, toReplace) {
+		for i := min; i <= max; i++ {
+			unparsedURL := strings.Replace(urlArg, toReplace, strconv.Itoa(i), 1)
+			parsedURL, err := url.Parse(unparsedURL)
```

Comment:
URL parsing is duplicated here and in the else-part of ParseURLArgument(). Suggestion: First generate all URLs (in case there is no template, the "generation" is trivial), THEN do the URL parsing on all resulting URLs.